pkg/ring/incremental_transfer.go

```diff
@@ -0,0 +1,386 @@
+					fmt.Sprintf("failed to update consul when changing token %d to ACTIVE", token))
+			}
+		}
+
+		for addr, ranges := range workload {
+			pendingUnblocks = append(pendingUnblocks, struct {
+				addr string
+				rgs  []TokenRange
+			}{addr, ranges})
+		}
+	}
+
+	go func() {
+		if i.cfg.RangeUnblockDelay != time.Duration(0) {
+			<-time.After(i.cfg.RangeUnblockDelay)
+		}
+
+		ctx := context.Background()
+
+		for _, block := range pendingUnblocks {
+			err := i.incTransferer.UnblockRanges(ctx, block.rgs, block.addr)
+			if err != nil {
+				level.Error(util.Logger).Log(
+					"msg", "failed to unblock transferred ranges",
+					"target_addr", block.addr,
+					"ranges", PrintableRanges(block.rgs),
+					"err", err,
+				)
+			}
+		}
+	}()
+
+	return nil
+}
+
+// leaveIncrementalTransfer will attempt to incrementally send chunks to
+// neighboring lifecyclers that should contain data for token ranges the
+// leaving lifecycler will no longer receive writes for.
+func (i *Lifecycler) leaveIncrementalTransfer(ctx context.Context) error {
+	// Make sure all tokens are set to leaving, even when we
+	// fail.
+	defer func() {
+		i.setTokens(nil)
+		i.setState(LEAVING)
+		i.updateConsul(ctx)
+
+		remainingTokens := i.incTransferer.StreamTokens()
+		if len(remainingTokens) > 0 {
+			level.Warn(util.Logger).Log(
+				"msg", "not all tokens transferred out",
+				"streams_remaining", len(remainingTokens),
+			)
+
+			printTokens := remainingTokens
+			if len(printTokens) > 20 {
+				printTokens = printTokens[:20]
+			}
+
+			level.Debug(util.Logger).Log(
+				"msg", "non-transferred tokens",
+				"tokens", printTokens,
+			)
+		}
+	}()
+
+	r, err := i.getDenormalisedRing(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to read ring: %v", err)
+	}
+
+	replicationFactor := i.cfg.RingConfig.ReplicationFactor
+	if active := r.FindIngestersByState(ACTIVE); len(active) <= replicationFactor {
+		return fmt.Errorf("not transferring out; number of ingesters less than or equal to replication factor")
+	}
+
+	success := true
+
+	i.setTransitioningTokens(i.getTokens())
+	tokens := i.getTransitioningTokens()
+
+	var wg sync.WaitGroup
+
+	for tokidx, token := range tokens {
+		// For findTransferWorkload to find the full list of tokens to transfer,
+		// we first set the token to LEAVING locally and refresh it in our
+		// copy of the ring.
+		i.stateMtx.Lock()
+		token.State = LEAVING
+		tokens[tokidx] = token
+		i.stateMtx.Unlock()
+
+		r.RefreshTokenState(i.ID, tokens, false)
+
+		workload, ok := i.findTransferWorkload(r, token)
+		if !ok {
+			success = false
+		}
+
+		for addr, ranges := range workload {
+			wg.Add(1)
+			go func(addr string, ranges []TokenRange) {
+				err := i.incTransferer.BlockRanges(ctx, ranges, addr)
+				if err == nil {
+					// Also block the range locally. Will not be unblocked.
+					err = i.incTransferer.BlockRanges(ctx, ranges, "")
+				}
+
+				if err != nil {
+					level.Error(util.Logger).Log(
```

Comment:
Please don't split these log messages to 6 lines (here, and elsewhere in this method). It's just a single log message, but this way of splitting it adds lot of visual clutter and it makes it difficult to quickly see the flow of the code.

Reply:
I need to think about this one. I personally don't find it hard to read and it scans very similarly to how structs are instantiated. I get the argument that it's just a log line, but `level.Error(util.Logger).Log("msg", "failed to send chunks", "target_addr", addr, "ranges", PrintableRanges(ranges), "err", err)` would also bring attention to it and disrupt the flow of reading just by its sheer length. 

@gouthamve Is there a style standard for Cortex w/r/t line breaks?

Reply:
This is what I talk about ... 15 lines of visual clutter for what’s effectively 3 log messages. <img width="1194" src="https://user-images.githubusercontent.com/895919/67987599-26227b80-fc2e-11e9-899c-55f381cbbd81.png">

Reply:
My guess would be: https://github.com/golang/go/wiki/CodeReviewComments